File: src/lib/model-sizing.ts
// Model performance sizing — pure, dependency-free, and shared by the server
// report, the browser (live recalculation while the sliders move) and, as a
// port, the iOS app (ModelSizing.swift). Keep the three in step.
//
// Every historical leg is re-sized from the model settings: backing = equity ×
// share traded × margin available, split equally across the legs of a basket
// (times the basket's allocation scale), side toggles removing legs entirely.
// Per-contract economics (credit, P&L, margin) come from the published leg and
// scale with the contract count.

export interface SizingSettings {
  modelEquity: number;
  accountTradedPct: number;
  marginAvailablePct: number;
  sellCalls: boolean;
  sellPuts: boolean;
}

// One published leg with its settled outcome (pnl null while open).
export interface PerformanceLegSource {
  positionId: string;
  ticker: string;
  side: "call" | "put";
  strike: number;
  entryPrice: number;
  entryCredit: number;
  contracts: number;
  margin: number;
  credit: number;
  pnl: number | null;
  state: string | null;
  settledAt: string | null;
}

export interface PerformanceWeekSource {
  weekOf: string;
  slug: string;
  title: string;
  gsrs: number;
  cashNeeded: number;
  allocationScale: number;
  legs: PerformanceLegSource[];
}

export interface SettledLeg {
  positionId: string;
  ticker: string;
  side: "call" | "put";
  strike: number;
  entryCredit: number;
  contracts: number;
  margin: number;
  pnl: number;
  state: string;
  settledAt: string;
}

export interface WeeklyPerformance {
  weekOf: string;
  slug: string;
  title: string;
  gsrs: number;
  legs: number;
  settledLegs: number;
  wins: number;
  losses: number;
  pnl: number;
  margin: number;
  cashNeeded: number;
  credit: number;
  romPct: number | null;
  worstLeg: SettledLeg | null;
  complete: boolean;
}

export interface PerformanceStats {
  totalPnl: number;
  completeWeeks: number;
  winningWeeks: number;
  losingWeeks: number;
  avgWeeklyPnl: number;
  avgWinningWeek: number;
  avgLosingWeek: number;
  bestWeek: number;
  worstWeek: number;
  legWinRatePct: number;
  settledLegs: number;
  maxDrawdown: number;
  worstLeg: SettledLeg | null;
}

// How the legs were sized for a report. "model": every historical leg is
// re-sized from the model settings, so changing them changes the whole track
// record. "published": the contracts each basket was published with.
export interface PerformanceBasis {
  sizing: "model" | "published";
  modelEquity: number | null;
  accountTradedPct: number | null;
  marginAvailablePct: number | null;
  sellCalls: boolean;
  sellPuts: boolean;
  backing: number | null;
}

export interface ModelPerformance {
  weeks: WeeklyPerformance[];
  cumulative: { weekOf: string; pnl: number; cumulative: number }[];
  stats: PerformanceStats;
  basis: PerformanceBasis;
}

export const SLIDER_RANGES = {
  accountTradedPct: { min: 0, max: 100, step: 1 },
  marginAvailablePct: { min: 100, max: 1000, step: 25 },
} as const;

// Dollars of strike-or-spot notional the settings back. Mirrors
// shared/broker-settings.mjs sizingBacking without importing it, so this file
// stays safe to ship to the browser.
export function modelBacking(settings: SizingSettings) {
  const traded = Number(settings.accountTradedPct);
  const margin = Number(settings.marginAvailablePct);
  return Math.max(0, Number(settings.modelEquity) || 0) * (Number.isFinite(traded) ? traded : 100) / 100 * (Number.isFinite(margin) ? margin : 100) / 100;
}

// Re-size one leg under a sizing policy. Per-contract economics come from the
// published leg; the contract count is what the policy would have bought.
export function resizeLeg(leg: { entryPrice: number; strike: number; contracts: number; margin: number; credit: number; pnl: number | null }, perTradeBacking: number) {
  const unit = Math.max(leg.entryPrice, leg.strike) * 100;
  const contracts = unit > 0 && perTradeBacking > 0 ? Math.floor(perTradeBacking / unit) : 0;
  const ratio = leg.contracts > 0 ? contracts / leg.contracts : 0;
  return {
    contracts,
    margin: Math.round(leg.margin * ratio),
    credit: Math.round(leg.credit * ratio),
    pnl: leg.pnl == null ? null : leg.pnl * ratio,
  };
}

export function sideEnabled(side: "call" | "put", settings: SizingSettings) {
  return side === "call" ? settings.sellCalls : settings.sellPuts;
}

// Settled-performance report across every basket under a sizing policy
// (null = the published contracts). "Modeled" throughout: entries at the
// recommended credit, held to expiry or exited on a radar signal — the raw
// quality of the recommendations, not a record of executed trades.
export function computeModelPerformance(source: PerformanceWeekSource[], model: SizingSettings | null): ModelPerformance {
  const backing = model ? modelBacking(model) : null;
  const basis: PerformanceBasis = model
    ? { sizing: "model", modelEquity: model.modelEquity, accountTradedPct: model.accountTradedPct, marginAvailablePct: model.marginAvailablePct, sellCalls: model.sellCalls, sellPuts: model.sellPuts, backing }
    : { sizing: "published", modelEquity: null, accountTradedPct: null, marginAvailablePct: null, sellCalls: true, sellPuts: true, backing: null };

  const weeks: WeeklyPerformance[] = [];
  for (const week of source) {
    // Side toggles remove legs from the model entirely; the remaining legs share the backing.
    const legs = model ? week.legs.filter((leg) => sideEnabled(leg.side, model)) : week.legs;
    if (legs.length === 0) continue;
    const perTrade = backing != null ? (backing * (week.allocationScale || 1)) / legs.length : null;

    let pnl = 0, wins = 0, losses = 0, settled = 0, margin = 0, credit = 0;
    let worstLeg: SettledLeg | null = null;
    for (const leg of legs) {
      const sized = perTrade != null ? resizeLeg(leg, perTrade) : leg;
      margin += sized.margin;
      credit += sized.credit;
      if (sized.pnl == null || leg.state == null || leg.settledAt == null) continue;
      settled += 1;
      pnl += sized.pnl;
      if (sized.pnl >= 0) wins += 1;
      else losses += 1;
      const settledLeg: SettledLeg = {
        positionId: leg.positionId, ticker: leg.ticker, side: leg.side, strike: leg.strike, entryCredit: leg.entryCredit,
        contracts: sized.contracts, margin: sized.margin, pnl: Math.round(sized.pnl), state: leg.state, settledAt: leg.settledAt,
      };
      if (!worstLeg || settledLeg.pnl < worstLeg.pnl) worstLeg = settledLeg;
    }

    weeks.push({
      weekOf: week.weekOf, slug: week.slug, title: week.title, gsrs: week.gsrs,
      legs: legs.length, settledLegs: settled, wins, losses, pnl: Math.round(pnl), margin, cashNeeded: week.cashNeeded, credit,
      romPct: margin > 0 ? +((pnl / margin) * 100).toFixed(2) : null,
      worstLeg, complete: settled === legs.length && settled > 0,
    });
  }
  weeks.sort((a, b) => a.weekOf.localeCompare(b.weekOf));

  const completeWeeks = weeks.filter((w) => w.complete);
  let running = 0, peak = 0, maxDrawdown = 0;
  const cumulative = completeWeeks.map((w) => {
    running += w.pnl;
    peak = Math.max(peak, running);
    maxDrawdown = Math.min(maxDrawdown, running - peak);
    return { weekOf: w.weekOf, pnl: w.pnl, cumulative: running };
  });

  const winningWeeks = completeWeeks.filter((w) => w.pnl >= 0);
  const losingWeeks = completeWeeks.filter((w) => w.pnl < 0);
  const totalWins = completeWeeks.reduce((a, w) => a + w.wins, 0);
  const totalLegs = completeWeeks.reduce((a, w) => a + w.settledLegs, 0);
  const allWorst = completeWeeks.map((w) => w.worstLeg).filter((l): l is SettledLeg => l != null).sort((a, b) => a.pnl - b.pnl);
  const sum = (list: WeeklyPerformance[]) => list.reduce((a, w) => a + w.pnl, 0);

  const stats: PerformanceStats = {
    totalPnl: Math.round(sum(completeWeeks)),
    completeWeeks: completeWeeks.length,
    winningWeeks: winningWeeks.length,
    losingWeeks: losingWeeks.length,
    avgWeeklyPnl: completeWeeks.length ? Math.round(sum(completeWeeks) / completeWeeks.length) : 0,
    avgWinningWeek: winningWeeks.length ? Math.round(sum(winningWeeks) / winningWeeks.length) : 0,
    avgLosingWeek: losingWeeks.length ? Math.round(sum(losingWeeks) / losingWeeks.length) : 0,
    bestWeek: completeWeeks.length ? Math.max(...completeWeeks.map((w) => w.pnl)) : 0,
    worstWeek: completeWeeks.length ? Math.min(...completeWeeks.map((w) => w.pnl)) : 0,
    legWinRatePct: totalLegs ? +((totalWins / totalLegs) * 100).toFixed(1) : 0,
    settledLegs: totalLegs,
    maxDrawdown: Math.round(maxDrawdown),
    worstLeg: allWorst[0] ?? null,
  };

  return { weeks, cumulative, stats, basis };
}

File: ios/Sources/ModelSizing.swift
import Foundation

// Port of src/lib/model-sizing.ts: the model track record re-sized from the
// model settings, computed on the phone as the sliders move. Keep the two in
// step — the server returns the same numbers for the saved settings.

struct PerformanceLegSource: Codable, Identifiable {
    var id: String { positionId }
    let positionId: String
    let ticker: String
    let side: String
    let strike: Double
    let entryPrice: Double
    let entryCredit: Double
    let contracts: Int
    let margin: Double
    let credit: Double
    let pnl: Double?
    let state: String?
    let settledAt: String?
    let expiryPrice: Double?
}

struct PerformanceWeekSource: Codable, Identifiable {
    var id: String { slug }
    let weekOf: String
    let slug: String
    let title: String
    let gsrs: Double
    let cashNeeded: Double
    let allocationScale: Double
    let legs: [PerformanceLegSource]
}

struct ModelPerformance {
    var weeks: [WeekRow]
    var cumulative: [CumulativePoint]
    var stats: Stats
    var basis: PerformanceBasis
}

enum ModelSizing {
    static let tradedRange: ClosedRange<Double> = 0...100
    static let tradedStep = 1.0
    static let marginRange: ClosedRange<Double> = 100...1000
    static let marginStep = 25.0

    // Dollars of strike-or-spot notional the settings back.
    static func backing(_ s: ModelSettings) -> Double {
        max(0, s.modelEquity) * s.accountTradedPct / 100 * s.marginAvailablePct / 100
    }

    struct SizedLeg { let contracts: Int; let margin: Double; let credit: Double; let pnl: Double? }

    static func resize(_ leg: PerformanceLegSource, perTrade: Double) -> SizedLeg {
        let unit = max(leg.entryPrice, leg.strike) * 100
        let contracts = unit > 0 && perTrade > 0 ? Int((perTrade / unit).rounded(.down)) : 0
        let ratio = leg.contracts > 0 ? Double(contracts) / Double(leg.contracts) : 0
        return SizedLeg(contracts: contracts, margin: (leg.margin * ratio).rounded(), credit: (leg.credit * ratio).rounded(), pnl: leg.pnl.map { $0 * ratio })
    }

    struct LegResult { let ticker: String; let side: String; let strike: Double; let entryPrice: Double; let expiryPrice: Double?; let pnl: Double; let credit: Double; let margin: Double; let cushionPct: Double }

    static func compute(_ source: [PerformanceWeekSource], model: ModelSettings) -> ModelPerformance {
        let backing = backing(model)
        var weeks: [WeekRow] = []
        var legsByWeek: [String: [LegResult]] = [:]
        for week in source {
            let legs = week.legs.filter { $0.side == "call" ? model.sellCalls : model.sellPuts }
            if legs.isEmpty { continue }
            let perTrade = backing * (week.allocationScale > 0 ? week.allocationScale : 1) / Double(legs.count)
            var pnl = 0.0, margin = 0.0, credit = 0.0
            var wins = 0, losses = 0, settled = 0
            var results: [LegResult] = []
            for leg in legs {
                let sized = resize(leg, perTrade: perTrade)
                margin += sized.margin
                credit += sized.credit
                guard let legPnl = sized.pnl, leg.state != nil, leg.settledAt != nil else { continue }
                settled += 1
                pnl += legPnl
                if legPnl >= 0 { wins += 1 } else { losses += 1 }
                let cushion = leg.side == "call" ? leg.strike - leg.entryPrice : leg.entryPrice - leg.strike
                results.append(LegResult(ticker: leg.ticker, side: leg.side, strike: leg.strike, entryPrice: leg.entryPrice, expiryPrice: leg.expiryPrice,
                                         pnl: legPnl.rounded(), credit: leg.entryCredit * 100 * Double(sized.contracts), margin: sized.margin,
                                         cushionPct: leg.entryPrice > 0 ? cushion / leg.entryPrice * 100 : 0))
            }
            let complete = settled == legs.count && settled > 0
            let worst = results.min { $0.pnl < $1.pnl }
            weeks.append(WeekRow(weekOf: week.weekOf, slug: week.slug, title: week.title, gsrs: week.gsrs, legs: legs.count, settledLegs: settled,
                                 wins: wins, losses: losses, pnl: pnl.rounded(), margin: Int(margin), credit: Int(credit),
                                 romPct: margin > 0 ? (pnl / margin * 100 * 100).rounded() / 100 : nil, complete: complete,
                                 worstLeg: worst.map { Stats.WorstLeg(ticker: $0.ticker, side: $0.side, pnl: $0.pnl, strike: $0.strike, entryPrice: $0.entryPrice, expiryPrice: $0.expiryPrice) }))
            if complete { legsByWeek[week.slug] = results }
        }
        weeks.sort { $0.weekOf < $1.weekOf }

        let complete = weeks.filter(\.complete)
        var running = 0.0, peak = 0.0, maxDrawdown = 0.0
        let cumulative = complete.map { w -> CumulativePoint in
            running += w.pnl
            peak = max(peak, running)
            maxDrawdown = min(maxDrawdown, running - peak)
            return CumulativePoint(weekOf: w.weekOf, pnl: w.pnl, cumulative: running)
        }
        let winning = complete.filter { $0.pnl >= 0 }
        let losing = complete.filter { $0.pnl < 0 }
        let totalWins = complete.reduce(0) { $0 + $1.wins }
        let totalLegs = complete.reduce(0) { $0 + $1.settledLegs }
        func sum(_ list: [WeekRow]) -> Double { list.reduce(0) { $0 + $1.pnl } }
        func avg(_ list: [WeekRow]) -> Double { list.isEmpty ? 0 : (sum(list) / Double(list.count)).rounded() }
        func mean(_ list: [Double]) -> Double? { list.isEmpty ? nil : list.reduce(0, +) / Double(list.count) }
        let allLegs = complete.flatMap { legsByWeek[$0.slug] ?? [] }
        let grossWins = sum(winning), grossLosses = -sum(losing)
        let legCredit = allLegs.reduce(0) { $0 + $1.credit }, legMargin = allLegs.reduce(0) { $0 + $1.margin }, legPnl = allLegs.reduce(0) { $0 + $1.pnl }
        let weeklyMean = mean(complete.map(\.pnl)) ?? 0
        let weeklySd = complete.count >= 3 ? (complete.reduce(0) { $0 + ($1.pnl - weeklyMean) * ($1.pnl - weeklyMean) } / Double(complete.count - 1)).squareRoot() : 0
        var longestLosing = 0, run = 0
        for w in complete { run = w.pnl < 0 ? run + 1 : 0; longestLosing = max(longestLosing, run) }
        var currentStreak = 0
        for w in complete.reversed() {
            let sign = w.pnl >= 0 ? 1 : -1
            if currentStreak == 0 { currentStreak = sign } else if (currentStreak > 0) == (sign > 0) { currentStreak += sign } else { break }
        }
        func side(_ s: String) -> SidePerformance {
            let legs = allLegs.filter { $0.side == s }
            let wins = legs.filter { $0.pnl >= 0 }.count
            return SidePerformance(legs: legs.count, wins: wins, pnl: legs.reduce(0) { $0 + $1.pnl }.rounded(), credit: legs.reduce(0) { $0 + $1.credit }.rounded(),
                                   winRatePct: legs.isEmpty ? 0 : (Double(wins) / Double(legs.count) * 1000).rounded() / 10)
        }
        func r1(_ v: Double?) -> Double? { v.map { ($0 * 10).rounded() / 10 } }
        let worstOverall = allLegs.min { $0.pnl < $1.pnl }
        let stats = Stats(
            totalPnl: sum(complete).rounded(),
            completeWeeks: complete.count, winningWeeks: winning.count, losingWeeks: losing.count,
            avgWeeklyPnl: avg(complete), avgWinningWeek: avg(winning), avgLosingWeek: avg(losing),
            bestWeek: complete.map(\.pnl).max() ?? 0, worstWeek: complete.map(\.pnl).min() ?? 0,
            legWinRatePct: totalLegs > 0 ? (Double(totalWins) / Double(totalLegs) * 1000).rounded() / 10 : 0,
            settledLegs: totalLegs, maxDrawdown: maxDrawdown.rounded(),
            worstLeg: worstOverall.map { Stats.WorstLeg(ticker: $0.ticker, side: $0.side, pnl: $0.pnl, strike: $0.strike, entryPrice: $0.entryPrice, expiryPrice: $0.expiryPrice) },
            profitFactor: grossLosses > 0 ? (grossWins / grossLosses * 100).rounded() / 100 : nil,
            expectancyPerLeg: allLegs.isEmpty ? 0 : (legPnl / Double(allLegs.count)).rounded(),
            creditCapturePct: legCredit > 0 ? (legPnl / legCredit * 1000).rounded() / 10 : nil,
            returnOnMarginPct: legMargin > 0 ? (legPnl / legMargin * 10000).rounded() / 100 : nil,
            sharpe: weeklySd > 0 ? (weeklyMean / weeklySd * 52.0.squareRoot() * 100).rounded() / 100 : nil,
            longestLosingStreak: longestLosing, currentStreak: currentStreak,
            avgCushionPct: r1(mean(allLegs.map(\.cushionPct))),
            avgCushionWinnersPct: r1(mean(allLegs.filter { $0.pnl >= 0 }.map(\.cushionPct))),
            avgCushionLosersPct: r1(mean(allLegs.filter { $0.pnl < 0 }.map(\.cushionPct))),
            calls: side("call"), puts: side("put"))
        let basis = PerformanceBasis(sizing: "model", modelEquity: model.modelEquity, accountTradedPct: model.accountTradedPct,
                                     marginAvailablePct: model.marginAvailablePct, sellCalls: model.sellCalls, sellPuts: model.sellPuts)
        return ModelPerformance(weeks: weeks, cumulative: cumulative, stats: stats, basis: basis)
    }
}
